docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+
+```javascript
+function promptPromise(message) {
+  return new Promise(function(resolve, reject) {
+    var result = window.prompt(message);
+    if (result != null) {
+      resolve(result);
+    } else {
+      reject();
+    }
+  });
+}
+
+document.addEventListener('DOMContentLoaded', function() {
+  var button = document.getElementById('action');
+  var output = document.getElementById('prompt');
+
+  button.addEventListener('click', function() {
+    promptPromise('What is your name?')
+      .then(function(name) {
+        output.innerHTML = '' + name;
+      })
+      .catch(function() {
+        output.innerHTML = '¯\\_(ツ)_/¯';
+      });
+  });
+});
+```
+
+[Run example on JSBin](http://jsbin.com/kowama/edit?js,output)
+
+This doesn't add much much using `window.prompt`; however, one advantage is the
+API that promises provide. In the case where we call `promisePrompt(…)` we can
+easily react to the result of the dialog without having to worry about how it is
+implemented. In our example we've implemented the `window.prompt` but our call
+to `promisePrompt()` doesn't care. This makes a change to an *asynchronous*
+dialog a little more future proof.
+
+To drive home the synchronous nature of the `window.prompt` notice that the time
+stops ticking when the prompt dialog is displayed. Lets fix that by making our
+own prompt. Since our dialog is just DOM manipulation the page won't be blocked
+while waiting for user input.
+
+First add the prompt dialog to the HTML:
+
+```html
+<style type="text/css">
+  #dialog {
+    width:      200px;
+    margin:     auto;
+    padding:    10px;
+    border:     thin solid black;
+    background: lightgreen;
+  }
+  .hidden {
+    display: none;
+  }
+</style>
+<div id="dialog" class="hidden">
+  <div class="message">foobar</div>
+  <input type="text">
+  <div>
+    <button class="ok">Ok</button>
+    <button class="cancel">Cancel</button>
+  </div>
+</div>
+```
+
+We will want to keep the same API so our change will be only to the
+`promisePrompt`. It will find the dialog DOM elements, attach events to the
+elements, show the dialog box, return a promise that is resolved based on the
+attached events, and finally detaches the events and cleans up after itself
+(hiding the dialog box for another use later).
+
+```javascript
+function promptPromise(message) {
+  var dialog       = document.getElementById('dialog');
+  var input        = dialog.querySelector('input');
+  var okButton     = dialog.querySelector('button.ok');
+  var cancelButton = dialog.querySelector('button.cancel');
+  var resolver, rejector;
+
+  dialog.querySelector('.message').innerHTML = '' + message;
+  dialog.className = '';
+
+  function handleOk() {
+    resolver(input.value);
+  }
+
+  function handleCancel() {
+    rejector();
+  }
+
+  okButton.addEventListener('click', handleOk);
+  cancelButton.addEventListener('click', handleCancel);
+
+  return new Promise(function(resolve, reject) {
```

Comment:
I see your point. I'll rework this using the button click as promise resolvers that trigger the dialog promise resolve/reject. Also I think I can avoid the problem with the handlers by attaching one click event to the dialog itself and use event delegation to tell which button is clicked.

Will experiment. Apologies for the slow responses; I just had a baby which has occupied much of my cognitive functioning.